syntax="proto2";
package sisyphus.Raft;

enum OpCode {
    kRequestVote = 1;
    kAppendEntries = 2;
    kInstallSnapshot = 3;
};

enum EntryType {
    kUnknown = 0;
    kNormal = 1;
    kConfigure = 2;
    kNoop = 3;
};

message Entry {

    optional EntryType type = 1;

    required uint64 term = 2;  // the entry was created
    optional uint64 index = 3; // internal for log store

    optional bytes data = 4;
    optional bytes context = 5;
};

message RequestVoteOps {

    message Request {
        required uint64 server_id = 1;
        required uint64 term = 2; // caller's term
        required uint64 last_log_term = 3;
        required uint64 last_log_index = 4;
    }

    message Response {
        required uint64 term = 1;    // callee's term
        required bool   granted = 2;
        optional bool   log_ok = 3;  // hint for caller's log update to date
    }
}


message AppendEntriesOps {

    message Request {
        required uint64 server_id = 1; // id of caller(leader), send to calle's
        required uint64 term = 2;
        required uint64 pre_log_index = 3;
        required uint64 pre_log_term = 4;

        repeated Entry  entries = 5;
        required uint64 commit_index = 6;
    }

    message Response {
        required uint64 term = 1;
        required bool   success = 2;

        optional uint64 last_log_index = 3;
    }
}

message InstallSnapshotOps {

    message Request {
        required uint64 server_id = 1; // id of caller(leader), send to calle's
        required uint64 term = 2;

        required uint64 last_snapshot_index = 3;

        required uint64 bytes_offset = 5;
        required bytes  data = 6;
        required bool   done = 7;
    }

    message Response {
        required uint64 term = 1;
        optional uint64 bytes_stored = 2;
    }
}

